File: frontend/src/lib/schemas/source-health.ts
import { z } from "zod";

export const sourceHealthSummarySchema = z.object({
  total_sources: z.number().int(),
  indexed: z.number().int(),
  failed_indexing: z.number().int(),
  pending: z.number().int(),
  ocr_required: z.number().int(),
  ocr_low_confidence: z.number().int(),
  table_extraction_warnings: z.number().int(),
  missing_metadata: z.number().int(),
  stale: z.number().int(),
  deprecated: z.number().int(),
  expired: z.number().int(),
  unreviewed: z.number().int(),
  needs_review: z.number().int(),
  generated_at: z.string(),
});

export const sourceStatusCountSchema = z.object({
  status: z.string(),
  count: z.number().int(),
});

export const indexingFailurePointSchema = z.object({
  date: z.string(),
  failed_count: z.number().int(),
});

export const staleByCollectionSchema = z.object({
  collection_id: z.string().nullable(),
  collection_name: z.string(),
  stale_count: z.number().int(),
});

export const ocrQualityCountSchema = z.object({
  ocr_quality_status: z.string(),
  count: z.number().int(),
});

export const reviewNeedsByOwnerSchema = z.object({
  owner_id: z.string().nullable(),
  owner_name: z.string(),
  needs_review_count: z.number().int(),
});

export const connectorFreshnessSchema = z.object({
  connection_id: z.string(),
  connector_name: z.string(),
  provider_key: z.string().nullable(),
  last_successful_sync_at: z.string().nullable(),
  days_since_last_sync: z.number().int().nullable(),
  status: z.string(),
});

export const sourceHealthChartsSchema = z.object({
  status_distribution: z.array(sourceStatusCountSchema),
  indexing_failures: z.array(indexingFailurePointSchema),
  stale_by_collection: z.array(staleByCollectionSchema),
  ocr_quality_distribution: z.array(ocrQualityCountSchema),
  review_needs_by_owner: z.array(reviewNeedsByOwnerSchema),
  connector_freshness: z.array(connectorFreshnessSchema),
  generated_at: z.string(),
});

export const sourceTypeSchema = z.enum(["file", "connector", "collection"]);
export const freshnessSchema = z.enum(["fresh", "stale", "expired"]);

export const sourceHealthRowSchema = z.object({
  source_type: sourceTypeSchema,
  source_id: z.string(),
  source_name: z.string(),
  connector_name: z.string().nullable().optional(),
  collection_id: z.string().nullable().optional(),
  collection_name: z.string().nullable().optional(),
  owner_id: z.string().nullable().optional(),
  owner_name: z.string().nullable().optional(),
  status: z.string(),
  last_indexed_at: z.string().nullable().optional(),
  last_updated_at: z.string().nullable().optional(),
  freshness: freshnessSchema,
  trust_status: z.string().nullable().optional(),
  ocr_quality: z.string().nullable().optional(),
  review_status: z.string().nullable().optional(),
  graph_indexed: z.string().nullable().optional(),
  missing_metadata: z.boolean(),
  error_message: z.string().nullable().optional(),
  available_actions: z.array(z.string()),
});

export const sourceHealthListSchema = z.object({
  rows: z.array(sourceHealthRowSchema),
  total: z.number().int(),
  page: z.number().int(),
  page_size: z.number().int(),
});

export const tableWarningItemSchema = z.object({
  chunk_id: z.string(),
  page_number: z.number().int().nullable(),
  confidence: z.number().nullable(),
  reason: z.string(),
});

export const sourceHealthErrorDetailSchema = z.object({
  source_type: sourceTypeSchema,
  source_id: z.string(),
  source_name: z.string(),
  status: z.string(),
  error_message: z.string().nullable().optional(),
  extraction_warnings: z.array(z.string()),
  ocr_quality_status: z.string().nullable().optional(),
  ocr_avg_confidence: z.number().nullable().optional(),
  table_warnings: z.array(tableWarningItemSchema),
});

export type SourceHealthSummary = z.infer<typeof sourceHealthSummarySchema>;
export type SourceHealthCharts = z.infer<typeof sourceHealthChartsSchema>;
export type SourceType = z.infer<typeof sourceTypeSchema>;
export type Freshness = z.infer<typeof freshnessSchema>;
export type SourceHealthRow = z.infer<typeof sourceHealthRowSchema>;
export type SourceHealthList = z.infer<typeof sourceHealthListSchema>;
export type TableWarningItem = z.infer<typeof tableWarningItemSchema>;
export type SourceHealthErrorDetail = z.infer<
  typeof sourceHealthErrorDetailSchema
>;

export type SourceHealthQuery = {
  source_type?: SourceType;
  status?: string;
  collection_id?: string;
  owner_id?: string;
  freshness?: Freshness;
  review_status?: string;
  ocr_quality?: string;
  graph_indexed?: "yes" | "no" | "failed";
  missing_metadata?: boolean;
  q?: string;
  page?: number;
  page_size?: number;
};

File: backend/app/domains/admin/schemas/source_health.py
"""Pydantic schemas for the admin source health dashboard API (F352).

Returned by GET /admin/source-health/*. Aggregates existing indexing,
OCR, trust/review, connector sync, and metadata-completeness signals
already tracked on Document/Collection/connector models into a single
health rollup. No new persisted fields are introduced by this feature.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

# Source types this dashboard understands. "file" and "connector" are both
# Document rows, split by Document.ingestion_source. "collection" is a
# Collection row summarizing its member documents. Graph-backed sources are
# NOT a fourth row type (that would double-count documents already listed as
# file/connector) — graph indexing status is instead surfaced as the
# `graph_indexed` facet on file/connector rows.
SourceType = str  # "file" | "connector" | "collection"
Freshness = str  # "fresh" | "stale" | "expired"


class SourceHealthSummaryResponse(BaseModel):
    """Org-scoped counters for the top-of-dashboard metric cards."""

    total_sources: int = 0
    indexed: int = 0
    failed_indexing: int = 0
    pending: int = 0
    ocr_required: int = 0
    ocr_low_confidence: int = 0
    table_extraction_warnings: int = 0
    missing_metadata: int = 0
    stale: int = 0
    deprecated: int = 0
    expired: int = 0
    unreviewed: int = 0
    needs_review: int = 0
    generated_at: datetime


class SourceStatusCount(BaseModel):
    status: str
    count: int


class IndexingFailurePoint(BaseModel):
    date: str
    failed_count: int


class StaleByCollection(BaseModel):
    collection_id: str | None
    collection_name: str
    stale_count: int


class OcrQualityCount(BaseModel):
    ocr_quality_status: str
    count: int


class ReviewNeedsByOwner(BaseModel):
    owner_id: str | None
    owner_name: str
    needs_review_count: int


class ConnectorFreshness(BaseModel):
    connection_id: str
    connector_name: str
    provider_key: str | None
    last_successful_sync_at: datetime | None
    days_since_last_sync: int | None
    status: str


class SourceHealthChartsResponse(BaseModel):
    status_distribution: list[SourceStatusCount] = Field(default_factory=list)
    indexing_failures: list[IndexingFailurePoint] = Field(default_factory=list)
    stale_by_collection: list[StaleByCollection] = Field(default_factory=list)
    ocr_quality_distribution: list[OcrQualityCount] = Field(default_factory=list)
    review_needs_by_owner: list[ReviewNeedsByOwner] = Field(default_factory=list)
    connector_freshness: list[ConnectorFreshness] = Field(default_factory=list)
    generated_at: datetime


class SourceHealthRow(BaseModel):
    source_type: SourceType
    source_id: str
    source_name: str
    connector_name: str | None = None
    collection_id: str | None = None
    collection_name: str | None = None
    owner_id: str | None = None
    owner_name: str | None = None
    status: str
    last_indexed_at: datetime | None = None
    last_updated_at: datetime | None = None
    freshness: Freshness
    # Raw trust_status (Document only; null for collections). Exposed so the
    # frontend can PATCH .../trust-status with the current value unchanged
    # when it only wants to touch review_owner_id/review_due_date.
    trust_status: str | None = None
    ocr_quality: str | None = None
    review_status: str | None = None
    graph_indexed: str | None = None
    missing_metadata: bool = False
    error_message: str | None = None
    available_actions: list[str] = Field(default_factory=list)


class SourceHealthListResponse(BaseModel):
    rows: list[SourceHealthRow] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    page_size: int = 25


class TableWarningItem(BaseModel):
    chunk_id: str
    page_number: int | None
    confidence: float | None
    reason: str


class SourceHealthErrorDetailResponse(BaseModel):
    source_type: SourceType
    source_id: str
    source_name: str
    status: str
    error_message: str | None = None
    extraction_warnings: list[str] = Field(default_factory=list)
    ocr_quality_status: str | None = None
    ocr_avg_confidence: float | None = None
    table_warnings: list[TableWarningItem] = Field(default_factory=list)
